File: backend/aiconsole/core/chat/locations.py
from typing import cast

from aiconsole.core.chat.actor_id import ActorId
from aiconsole.core.chat.types import (
    AICChat,
    AICChatOptions,
    AICMessage,
    AICMessageGroup,
    AICToolCall,
)
from aiconsole.core.code_running.code_interpreters.language import LanguageStr
from fastmutation.data_context import DataContext
from fastmutation.types import (
    AttributeRef,
    CollectionRef,
    ObjectRef,
    StringAttributeRef,
)


class AssetCollectionRef(CollectionRef):
    id: str = "assets"
    parent: None = None


class AssetRef(ObjectRef):
    parent: AssetCollectionRef

    def __init__(self, id: str, context: DataContext):
        super().__init__(id=id, parent_collection=AssetCollectionRef(context=context), context=context)


class ChatRef(ObjectRef[AICChat]):
    parent: AssetCollectionRef

    def __init__(self, id: str, context: DataContext):
        super().__init__(id=id, parent_collection=AssetCollectionRef(context=context), context=context)

    @property
    def message_groups(self):
        return MessageGroupsRef(parent=self, context=self.context)

    @property
    def is_analysis_in_progress(self):
        return AttributeRef[bool](object=self, name="is_analysis_in_progress", context=self.context)

    @property
    def chat_options(self):
        return AttributeRef[AICChatOptions](object=self, name="chat_options", context=self.context)


class MessageGroupsRef(CollectionRef[AICMessageGroup]):
    id: str = "message_groups"
    parent: ChatRef

    def __getitem__(self, id: str) -> "MessageGroupRef":
        return MessageGroupRef(id=id, parent_collection=self, context=self.context)


class MessageGroupRef(ObjectRef[AICMessageGroup]):
    parent_collection: MessageGroupsRef

    @property
    def messages(self):
        return MessagesRef(parent=self, context=self.context)

    @property
    def actor_id(self):
        return AttributeRef[ActorId](object=self, name="actor_id", context=self.context)

    @property
    def materials_ids(self):
        return AttributeRef[list[str]](object=self, name="materials_ids", context=self.context)

    @property
    def task(self):
        return StringAttributeRef(object=self, name="task", context=self.context)

    @property
    def analysis(self):
        return StringAttributeRef(object=self, name="analysis", context=self.context)


class MessagesRef(CollectionRef[AICMessage]):
    id: str = "messages"
    parent: MessageGroupRef

    def __getitem__(self, id: str) -> "MessageRef":
        return MessageRef(id=id, parent_collection=self, context=self.context)


class MessageRef(ObjectRef[AICMessage]):
    parent_collection: MessagesRef

    @property
    def tool_calls(self):
        return ToolCallsRef(parent=self, context=self.context)

    @property
    def content(self):
        return StringAttributeRef(object=self, name="content", context=self.context)

    @property
    def is_streaming(self):
        return AttributeRef[bool](object=self, name="is_streaming", context=self.context)


class ToolCallsRef(CollectionRef[AICToolCall]):
    id: str = "tool_calls"
    parent: MessageRef

    def __getitem__(self, id: str) -> "ToolCallRef":
        return ToolCallRef(parent_collection=self, id=id, context=self.context)

    # HANDLE DELETE

    # Remove message group if it's empty
    # if not message_location.message_group.messages:
    #     chat.message_groups = [group for group in chat.message_groups if group.id != message_location.message_group.id]

    # Remove message group if it's empty
    # if not tool_call.message_group.messages:
    #    chat.message_groups = [group for group in chat.message_groups if group.id != tool_call.message_group.id]

    # SET VSALUE

    # _handle_SetMessageGroupAgentIdMutation
    # if mutation.actor_id == "user":
    #     message_group.role = "user"
    # else:
    #    message_group.role = "assistant"
    #    message_group.agent_id = mutation.actor_id


class ToolCallRef(ObjectRef):
    parent_collection: ToolCallsRef

    @property
    def headline(self):
        return StringAttributeRef(object=self, name="headline", context=self.context)

    @property
    def language(self):
        return StringAttributeRef[LanguageStr | None](object=self, name="language", context=self.context)

    @property
    def is_executing(self):
        return AttributeRef[bool](object=self, name="is_executing", context=self.context)

    @property
    def is_successful(self):
        return AttributeRef[bool](object=self, name="is_successful", context=self.context)

    @property
    def is_streaming(self):
        return AttributeRef[bool](object=self, name="is_streaming", context=self.context)

    @property
    def output(self):
        return StringAttributeRef[str | None](object=self, name="output", context=self.context)

    @property
    def code(self):
        return StringAttributeRef(object=self, name="code", context=self.context)

    async def get(self) -> AICToolCall:
        return cast(AICToolCall, await super().get())

File: frontend/src/store/assets/locations.ts
import { AICChatOptions } from '@/types/assets/chatTypes';
import { AttributeRef, CollectionRef, ObjectRef, StringAttributeRef } from './types';
import { AICChat, AICMessage, AICMessageGroup, AICToolCall, Asset } from './constructors';
import { DataContext } from './DataContext';

// COLLECTION REFS
export class AssetsCollectionRef extends CollectionRef<Asset> {
  constructor(context: DataContext | null = null) {
    super('assets', null, context);
  }
}

export class MessagesGroupCollectionRef extends CollectionRef<AICMessageGroup> {
  constructor(id: string, context: DataContext | null = null) {
    super('message_groups', new ChatRef(id, context), context);
  }

  getById(id: string) {
    return new MessageGroupRef(id, this.context);
  }
}

export class MessagesCollectionRef extends CollectionRef<AICMessage> {
  constructor(id: string, context: DataContext | null = null) {
    super('messages', new MessageGroupRef(id, context), context);
  }

  getById(id: string) {
    return new MessageRef(id, this.context);
  }
}

export class ToolCallsCollectionRef extends CollectionRef<AICToolCall> {
  constructor(id: string, context: DataContext | null = null) {
    super('tool_calls', new ToolCallRef(id, context), context);
  }

  getById(id: string) {
    return new ToolCallRef(id, this.context);
  }
}
// OBJECT REFS
export class AssetRef extends ObjectRef<Asset> {
  constructor(id: string, context: DataContext | null = null) {
    super(id, new AssetsCollectionRef(context), context);
  }
}

export class ChatRef extends ObjectRef<AICChat> {
  constructor(id: string, context: DataContext | null = null) {
    super(id, new AssetsCollectionRef(context), context);
  }

  get messagesGroups() {
    return new MessagesGroupCollectionRef(this.id, this.context);
  }

  get isAnalysisInProgress() {
    return new AttributeRef<boolean>('is_analysis_in_progress', this, this.context);
  }

  get chat_options() {
    return new AttributeRef<AICChatOptions>('chat_options', this, this.context);
  }
}

export class MessageGroupRef extends ObjectRef<AICMessageGroup> {
  constructor(id: string, context: DataContext | null = null) {
    super(id, new MessagesGroupCollectionRef(id, context), context);
  }

  get messages() {
    return new MessagesCollectionRef(this.id, this.context);
  }

  get actor_id() {
    return new AttributeRef<string>('actor_id', this, this.context);
  }

  get materials_ids() {
    return new AttributeRef<string[]>('materials_ids', this, this.context);
  }

  get task() {
    return new StringAttributeRef('task', this, this.context);
  }

  get analysis() {
    return new StringAttributeRef('analysis', this, this.context);
  }
}

export class MessageRef extends ObjectRef<AICMessage> {
  constructor(id: string, context: DataContext | null = null) {
    super(id, new MessagesCollectionRef(id, context), context);
  }

  get tool_calls() {
    return new ToolCallsCollectionRef(this.id, this.context);
  }

  get content() {
    return new StringAttributeRef('content', this, this.context);
  }

  get is_streaming() {
    return new AttributeRef<boolean>('is_streaming', this, this.context);
  }
}

export class ToolCallRef extends ObjectRef<AICToolCall> {
  constructor(id: string, context: DataContext | null = null) {
    super(id, new ToolCallsCollectionRef(id, context), context);
  }

  get headline() {
    return new StringAttributeRef('headline', this, this.context);
  }

  get language() {
    return new StringAttributeRef('language', this, this.context);
  }

  get is_executing() {
    return new AttributeRef<boolean>('is_executing', this, this.context);
  }

  get is_successful() {
    return new AttributeRef<boolean>('is_successful', this, this.context);
  }

  get is_streaming() {
    return new AttributeRef<boolean>('is_streaming', this, this.context);
  }

  get output() {
    return new StringAttributeRef('output', this, this.context);
  }

  get code() {
    return new StringAttributeRef('code', this, this.context);
  }
}
